File: benchmark/2026-09-17-7677581(latest)/micro/src/go/mb08.go
// MB-08 - Strings: text construction plus five character-level passes per round.
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

// lcg is the Lehmer / MINSTD generator, frozen for every language in the suite.
type lcg struct {
	state int64
}

func (g *lcg) nextInt() int64 {
	g.state = (48271 * g.state) % 2147483647
	return g.state
}

// rhash is an order-sensitive rolling hash over an ASCII byte sequence.
func rhash(seq []byte) int64 {
	var h int64
	for _, c := range seq {
		h = (h*131 + int64(c)) % 1000000007
	}
	return h
}

type result struct {
	acc, length, cntAB, cntW int64
}

// workload is the entire measured body: it re-seeds the generator and rebuilds
// the text, as methodology 06 section 5.2 requires of every steady iteration.
func workload() result {
	const (
		nw = 200000
		r  = 20
		q  = 1000000007
	)

	gen := lcg{state: 20268917}
	words := make([]string, nw)
	for w := 0; w < nw; w++ {
		l := 4 + int(gen.nextInt()%13) // length 4..16
		var sb strings.Builder
		for i := 0; i < l; i++ {
			sb.WriteByte(byte('a' + gen.nextInt()%26))
		}
		words[w] = sb.String()
	}
	text := []byte(strings.Join(words, " "))
	length := len(text)

	var acc, cntAB, cntW int64
	for round := 0; round < r; round++ {
		p := 7*round + 11 // anti-elimination; p <= 144 < length
		if text[p] == ' ' {
			text[p] = 'x'
		} else {
			text[p] = byte('a' + (int64(text[p])-'a'+1)%26)
		}

		h1 := rhash(text) // pass 1

		u := make([]byte, length) // pass 2: upper-case
		for i := 0; i < length; i++ {
			c := text[i]
			if c >= 97 && c <= 122 {
				u[i] = c - 32
			} else {
				u[i] = c
			}
		}
		h2 := rhash(u)

		v := make([]byte, length) // pass 3: reverse
		for i := 0; i < length; i++ {
			v[i] = text[length-1-i]
		}
		h3 := rhash(v)

		cntAB = 0 // pass 4: naive search
		for i := 0; i < length-1; i++ {
			if text[i] == 'a' && text[i+1] == 'b' {
				cntAB++
			}
		}

		// pass 5: word count on the language's own string type. Section 4.12(b)
		// pins Go to `string` built from the working buffer and accessed with
		// `for _, r := range s`; the construction is inside the timed round and
		// the string is never reused between rounds.
		s := string(text)
		cntW = 1
		for _, ch := range s {
			if ch == ' ' {
				cntW++
			}
		}

		for _, val := range [5]int64{h1, h2, h3, cntAB, cntW} {
			acc = (acc*31 + val) % q
		}
	}

	return result{acc: acc, length: int64(length), cntAB: cntAB, cntW: cntW}
}

// parseMode reads the section 5.2 program mode from argv[1] and the steady
// iteration count K from argv[2] (default 7).
func parseMode() (string, int) {
	mode := "once"
	if len(os.Args) > 1 {
		mode = os.Args[1]
	}
	k := 7
	if len(os.Args) > 2 {
		if v, err := strconv.Atoi(os.Args[2]); err == nil && v > 0 {
			k = v
		}
	}
	return mode, k
}

func main() {
	mode, k := parseMode()
	var r result
	if mode == "steady" {
		for i := 0; i < k; i++ {
			t0 := time.Now() // section 5.2 pins time.Now/time.Since for Go
			r = workload()
			elapsed := time.Since(t0).Nanoseconds()
			// os.Stdout is unbuffered in Go, so this Printf is an immediate write.
			fmt.Printf("ITER %d %d\n", i, elapsed)
		}
	} else {
		r = workload()
	}
	fmt.Printf("MB08 %d %d %d %d\n", r.acc, r.length, r.cntAB, r.cntW)
}

File: benchmark/2026-09-17-7677581(latest)/micro/src/rust/mb08.rs
// MB-08 - Strings: text construction plus five explicit character-level passes.
// Passes 1-4 run on the pinned mutable byte buffer (section 4.12(b): Vec<u8>);
// pass 5 runs on a native String built fresh each round, read with `.chars()`.
#![forbid(unsafe_code)]

use std::env;
use std::hint::black_box;
use std::io::Write;
use std::time::Instant;

/// Lehmer / MINSTD generator, frozen for every language in the suite.
struct Lcg {
    state: i64,
}

impl Lcg {
    fn new(seed: i64) -> Self {
        Lcg { state: seed }
    }

    fn next_int(&mut self) -> i64 {
        self.state = (48271 * self.state) % 2147483647;
        self.state
    }
}

/// Order-sensitive rolling hash over a byte sequence.
fn rhash(seq: &[u8]) -> i64 {
    let mut h: i64 = 0;
    for &c in seq {
        h = (h * 131 + c as i64) % 1000000007;
    }
    h
}

/// The whole workload body, build phase included (methodology 06 section 4, MB-08).
fn workload() -> (i64, i64, i64, i64) {
    const NW: usize = 200000;
    const R: usize = 20;
    const Q: i64 = 1000000007;

    let mut gen = Lcg::new(20268917);

    let mut words: Vec<String> = Vec::new();
    for _ in 0..NW {
        let l = 4 + (gen.next_int() % 13);
        let mut word = String::new();
        for _ in 0..l {
            word.push((b'a' + (gen.next_int() % 26) as u8) as char);
        }
        words.push(word);
    }
    let mut text = words.join(" ").into_bytes();
    let len = text.len();

    let mut acc: i64 = 0;
    let mut cnt_ab: i64 = 0;
    let mut cnt_w: i64 = 0;

    for r in 0..R {
        let p = 7 * r + 11; // anti-elimination; p <= 144 < len
        if text[p] == b' ' {
            text[p] = b'x';
        } else {
            text[p] = b'a' + (text[p] - b'a' + 1) % 26;
        }

        let h1 = rhash(&text); // pass 1

        let mut u = vec![0u8; len]; // pass 2: upper-case
        for i in 0..len {
            let c = text[i];
            u[i] = if c >= 97 && c <= 122 { c - 32 } else { c };
        }
        let h2 = rhash(&u);

        let mut v = vec![0u8; len]; // pass 3: reverse
        for i in 0..len {
            v[i] = text[len - 1 - i];
        }
        let h3 = rhash(&v);

        cnt_ab = 0; // pass 4: naive search
        for i in 0..len - 1 {
            if text[i] == b'a' && text[i + 1] == b'b' {
                cnt_ab += 1;
            }
        }

        // pass 5: word count on the language's own string type. The String is
        // constructed fresh from the working buffer inside the timed round and
        // read with Rust's ordinary in-order character API, `.chars()`
        // (section 4.12(b): Rust offers no integer character index).
        let s = String::from_utf8(text.clone()).expect("the working buffer is ASCII");
        cnt_w = 1;
        for c in s.chars() {
            if c == ' ' {
                cnt_w += 1;
            }
        }

        for value in [h1, h2, h3, cnt_ab, cnt_w] {
            acc = (acc * 31 + value) % Q;
        }
    }

    (acc, len as i64, cnt_ab, cnt_w)
}

fn main() {
    // Methodology 06 section 5.2 program modes: argv[1] in {once, steady}, argv[2] = K.
    let args: Vec<String> = env::args().collect();
    let mode = args.get(1).map(String::as_str).unwrap_or("once");
    let k: usize = args
        .get(2)
        .and_then(|s| s.parse::<usize>().ok())
        .unwrap_or(7)
        .max(1);

    let (acc, len, cnt_ab, cnt_w) = if mode == "steady" {
        let mut last = (0i64, 0i64, 0i64, 0i64);
        for i in 0..k {
            let start = Instant::now();
            last = black_box(workload());
            let elapsed_ns = start.elapsed().as_nanos();
            println!("ITER {} {}", i, elapsed_ns);
            std::io::stdout().flush().expect("stdout flush");
        }
        last
    } else {
        workload()
    };

    println!("MB08 {} {} {} {}", acc, len, cnt_ab, cnt_w);
}
